staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,617 @@
+	metrics.UpdateFlowControlRequestsExecuting(qs.config.Name, queue.RequestsExecuting)
+	request.Decision.SetLocked(DecisionExecute)
+	return request, ok
+}
+
+/// selectQueueLocked selects the minimum virtualFinish time from the set of queues
+// the starting queue is selected via roundrobin
+func (qs *queueSet) selectQueueLocked() *queue {
+	minVirtualFinish := math.Inf(1)
+	var minQueue *queue
+	var minIndex int
+	for range qs.queues {
+		qs.robinIndex = (qs.robinIndex + 1) % len(qs.queues)
+		queue := qs.queues[qs.robinIndex]
+		if len(queue.Requests) != 0 {
+			currentVirtualFinish := queue.GetVirtualFinish(0, qs.estimatedServiceTime)
+			if currentVirtualFinish < minVirtualFinish {
+				minVirtualFinish = currentVirtualFinish
+				minQueue = queue
+				minIndex = qs.robinIndex
+			}
+		}
+	}
+	// we set the round robin indexing to start at the chose queue
+	// for the next round.  This way the non-selected queues
+	// win in the case that the virtual finish times are the same
+	qs.robinIndex = minIndex
+	return minQueue
+}
+
+// finishRequestAndDispatchAsMuchAsPossible is a convenience method
+// which calls finishRequest for a given request and then dispatches
+// as many requests as possible.  This is all of what needs to be done
+// once a request finishes execution or is canceled.
+func (qs *queueSet) finishRequestAndDispatchAsMuchAsPossible(req *request) {
+	qs.lockAndSyncTime()
+	defer qs.lock.Unlock()
+
+	qs.finishRequestLocked(req)
+	qs.dispatchAsMuchAsPossibleLocked()
+}
+
+// finishRequestLocked is a callback that should be used when a
+// previously dispatched request has completed it's service.  This
+// callback updates important state in the queueSet
+func (qs *queueSet) finishRequestLocked(r *request) {
+	S := qs.clock.Since(r.StartTime).Seconds()
+
+	// When a request finishes being served, and the actual service time was S,
+	// the queue’s virtual start time is decremented by G - S.
+	r.Queue.VirtualStart -= qs.estimatedServiceTime - S
+
+	// request has finished, remove from requests executing
+	r.Queue.RequestsExecuting--
+
+	if klog.V(6) {
+		klog.Infof("QS(%s) at r=%s v=%.9fs: request %#+v %#+v finished, adjusted queue %d virtual start time to %.9fs due to service time %.9fs, queue will have %d waiting & %d executing", qs.config.Name, qs.clock.Now().Format(nsTimeFmt), qs.virtualTime, r.descr1, r.descr2, r.Queue.Index, r.Queue.VirtualStart, S, len(r.Queue.Requests), r.Queue.RequestsExecuting)
+	}
+
+	// Logic to remove quiesced queues
+	// >= as Index=25 is out of bounds for DesiredNum=25 [0...24]
+	if r.Queue.Index >= qs.config.DesiredNumQueues &&
+		len(r.Queue.Requests) == 0 &&
+		r.Queue.RequestsExecuting == 0 {
+		qs.queues = removeQueueAndUpdateIndexes(qs.queues, r.Queue.Index)
+
+		// decrement here to maintain the invariant that (qs.robinIndex+1) % numQueues
+		// is the index of the next queue after the one last dispatched from
+		if qs.robinIndex >= -r.Queue.Index {
```

Comment:
why negate r.Queue.Index ?

Reply:
Oh crud, is that typo still here?!

Reply:
Created #85257 to address this.

Reply:
Don't bother, I am making a bigger PR to fix many things.

Reply:
I LGTMd that, and am preparing another PR to fix all the other issues here.